ReSharper.FSharp/src/FSharp.TypeProviders.Protocol/src/TypeProvidersExternalProcessFactory.cs

```diff
@@ -16,49 +18,63 @@ public class TypeProvidersExternalProcessFactory
       [NotNull] ILogger logger,
       [NotNull] IShellLocks shellLocks,
-      [NotNull] ISolution solution,
-      [NotNull] ISolutionToolset toolset)
+      [NotNull] ISolutionToolset toolset,
+      [NotNull] OutputAssemblies outputAssemblies)
     {
       mySolutionProcessStartInfoPatcher = solutionProcessStartInfoPatcher;
       myLogger = logger;
       myShellLocks = shellLocks;
-      mySolution = solution;
       myToolset = toolset;
+      myOutputAssemblies = outputAssemblies;
     }
 
-    public TypeProvidersExternalProcess Create(Lifetime lifetime, bool isInternalMode)
+    public TypeProvidersExternalProcess Create(Lifetime lifetime,
+      [CanBeNull] string requestingProjectOutputAssemblyPath, bool isInternalMode)
     {
       var sdkVersion = myToolset.GetDotNetCoreToolset()?.Sdk?.Version;
 
       return new TypeProvidersExternalProcess(lifetime,
         myLogger,
         myShellLocks,
         mySolutionProcessStartInfoPatcher,
-        GetProcessRuntime(),
+        GetProcessRuntime(requestingProjectOutputAssemblyPath),
         sdkVersion,
         isInternalMode);
     }
 
-    private JetProcessRuntimeRequest GetProcessRuntime()
+    private JetProcessRuntimeRequest GetProcessRuntime([CanBeNull] string requestingProjectOutputAssemblyPath)
     {
-      var packageReferenceTracker = mySolution.GetComponent<NuGetPackageReferenceTracker>();
-
-      var installedPackages = packageReferenceTracker.GetAllInstalledPackages();
-      var containsLegacyCompiler = installedPackages.Any(x => x.PackageIdentity.Id == "FSharp.Compiler.Tools");
+      using var @lock = ReadLockCookie.Create();
 
       var buildTool = myToolset.GetBuildTool();
+
+      var runtimeType = buildTool!.UseDotNetCoreForLaunch
+        ? JetProcessRuntimeType.DotNetCore
+        : JetProcessRuntimeType.FullFramework;
+
+      if (requestingProjectOutputAssemblyPath != null)
+      {
+        var path = VirtualFileSystemPath.Parse(requestingProjectOutputAssemblyPath, InteractionContext.SolutionContext);
+        var project = myOutputAssemblies.TryGetProjectByOutputAssemblyLocation(path).NotNull();
+
+        foreach (var configuration in project.ProjectProperties.GetActiveConfigurations<IProjectConfiguration>())
+          if (configuration.PropertiesCollection.TryGetValue("FscToolExe", out var fscTool) &&
+              fscTool is "fsc.exe" or "fsharpc")
+            runtimeType = JetProcessRuntimeType.FullFramework;
+      }
+
       var mutator = MsBuildConnectionFactory.GetEnvironmentVariablesMutator(buildTool);
 
-      var runtimeRequest = buildTool!.UseDotNetCoreForLaunch && !containsLegacyCompiler
+      var runtimeRequest = runtimeType == JetProcessRuntimeType.DotNetCore
         ? JetProcessRuntimeRequest.CreateCore(mutator, true)
-        : JetProcessRuntimeRequest.CreateFramework(mutator: mutator);
+        : JetProcessRuntimeRequest.CreateFramework(mutator: mutator, useMono: !PlatformUtil.IsRunningUnderWindows);
```

Comment:
Isn't Mono automatically used when not on Windows? What runtime will it create if you don't use this flag?

Reply:
Indeed, there is logic that patches runtime to Mono when required.